### ISSUE TITLE
Links to content blocks

### ISSUE THREAD

**Issue by olayway** [flowershow/flowershow#62]
Links to a specific block (paragraph) within a given page, e.g. `[[roadmap#^dcf64c]]`.

See [obsidian docs](https://help.obsidian.md/How+to/Link+to+blocks) for reference.

## Design

**UPDATE 2023-12-12** from https://github.com/datopian/flowershow/issues/62#issuecomment-1843646839

To support these probably want a remark plugin or simple regex pattern match that search for `^six-digit-alphanumeric` and then converts that to html anchors in the output.

**Comment by rufuspollock** [comment]
@khalilcodes isn't this already supported in remark-wiki-link-plus (ie. you support `[[xyz#anchor-link]]`?

**Comment by rufuspollock** [comment]
Actually this may be tough depending on how obsidian inserts those anchors into the file. Need to research that.

**Comment by blacklightpy** [comment]
We can actually use custom anchors. Basically it's just some random anchors, you just have to look into the mentioned page and find the block containing the anchor.

**Comment by rufuspollock** [comment]
@blacklightpy how does obsidian encode the anchor link in markdown (and how does it insert the target anchor in markdown?)

**Comment by blacklightpy** [comment]
@rufuspollock So basically if we use the menu to create an anchor to a block, it will assign a random ID (6 digit alphanumeric) to that block by ending the block with `^ID`.

I can refer to it by [[Page To Link^ID]].

So I can manually name blocks by ending them with `^my-anchor` and I can refer to them by typing [[Page To Link^my-anchor]].

Spaces are not allowed in anchor names.

I also think anchors only have to be unique per page. So it's not stored in a database, but rather it's like simple HTML anchors.

**Comment by rufuspollock** [comment]
@blacklightpy thanks for the clear write-up.

So basically to implement you need to search for regex pattern `^six-digit-alphanumeric` and then convert that to html anchors in the processing.

**Comment by blacklightpy** [comment]
> @blacklightpy thanks for the clear write-up.
> 
> So basically to implement you need to search for regex pattern `^six-digit-alphanumeric` and then convert that to html anchors in the processing.

@rufuspollock No, you should search for `^any-length-alphanumeric-with-hypens` which appear at the end of a block, because custom anchors exist too.

**Comment by blacklightpy** [comment]
@rufuspollock More details (I just installed Obsidian after setting up my PC)

- **Important:** Actually it doesn't support underscores, only alphanumeric and hyphens.
- **Minor detail:** The anchor is not directly appended to the end of the block, but it is separated by a space character. That is, if the block ends with `done.`, the anchor should be `done. ^anchor`, not `done^anchor`. Although I don't think this will make an effective difference (exceptions being where the block may end with something like `x^2`)
- **Minor detail:** The anchor is technically the block ID. But block IDs are only generated when you specifically refer to them. For example, if you link to a block, then clear the generated block ID, and link to it again, the newly assigned block ID will be different.

**Comment by blacklightpy** [comment]
One more correction.. it's not [[Page to Link^anchor]], it is [[Page to Link#^anchor]].

Without the #, it will try to look for a page named "Page to Link^anchor".

# is for heading anchors and ^ is for blocks. So basically a block ID is implemented as a special type of heading anchor.
